File: amm/compiler/interface/src/main/java/ammonite/compiler/iface/CompilerLifecycleManager.java
package ammonite.compiler.iface;

import java.nio.file.Path;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.Map;

public abstract class CompilerLifecycleManager {
  public abstract void setRTCacheDir(Path dir);
  public abstract void setHeadFrame(Supplier<Frame> headFrame);
  public abstract void setDependencyCompleter(Supplier<Function<String, Map.Entry<Integer, String[]>>> dependencyCompleter);
  public abstract void setClassPathWhiteList(String[][] whiteList);
  public abstract void setInitialClassLoader(ClassLoader initialClassLoader);

  public abstract Compiler compiler();
  public abstract int compilationCount();

  public abstract Preprocessor preprocessor(String fileName);

  public abstract String userScalaVersion();

  public abstract void init();
  public abstract void forceInit();

  public abstract Map.Entry<Integer, Map.Entry<String[], String[]>> completions(int offset, String previousImports, String snippet);

  public abstract Compiler.Output compileClassOrNull(
    Preprocessor.Output processed,
    Logger printer,
    String fileName);

  public abstract void addToClasspath(Map.Entry<String, byte[]>[] classFiles);

  public abstract Object objPressy();
  public abstract void shutdownPressy();
}

File: amm/compiler/interface/src/main/scala/ammonite/compiler/iface/CompilerLifecycleManager.scala
package ammonite.compiler.iface

import java.nio.file.Path

import ammonite.util.{Frame, Printer}
import ammonite.util.Util.ClassFiles

abstract class CompilerLifecycleManager {
  def compiler: Compiler
  def compilationCount: Int

  def preprocess(fileName: String): Preprocessor

  def scalaVersion: String

  def outputDir: Option[Path]

  def init(force: Boolean = false): Unit

  def complete(
      offset: Int,
      previousImports: String,
      snippet: String
  ): (Int, Seq[String], Seq[String])

  def compileClass(
      processed: Preprocessor.Output,
      printer: Printer,
      fileName: String
  ): Option[Compiler.Output]

  def addToClasspath(classFiles: ClassFiles): Unit

  def shutdownPressy(): Unit
}
